fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,382 @@
+                Metrics aggregation function, optional.
+            fairness_metric : FairnessMetric, optional.
+                The metric to evaluate the local model of each client against the global model in order to
+                determine their adjustment weight for aggregation. Can be set to any default metric in
+                FairnessMetricType or set to use a custom metric. Optional, default is
+                FairnessMetric(FairnessMetricType.LOSS).
+            weight_step_size : float
+                The step size to determine the magnitude of change for the adjustment weight.
+                Optional, default is 0.2.
+        """
+        super().__init__(
+            fraction_fit=fraction_fit,
+            fraction_evaluate=fraction_evaluate,
+            min_fit_clients=min_fit_clients,
+            min_evaluate_clients=min_evaluate_clients,
+            min_available_clients=min_available_clients,
+            evaluate_fn=evaluate_fn,
+            on_fit_config_fn=on_fit_config_fn,
+            on_evaluate_config_fn=on_evaluate_config_fn,
+            accept_failures=accept_failures,
+            initial_parameters=initial_parameters,
+            fit_metrics_aggregation_fn=fit_metrics_aggregation_fn,
+            evaluate_metrics_aggregation_fn=evaluate_metrics_aggregation_fn,
+        )
+
+        if fairness_metric is None:
+            self.fairness_metric = FairnessMetric(FairnessMetricType.LOSS)
+        else:
+            self.fairness_metric = fairness_metric
+
+        self.weight_step_size = weight_step_size
+
+        self.train_metrics: Dict[str, Dict[str, Scalar]] = {}
+        self.evaluation_metrics: Dict[str, Dict[str, Scalar]] = {}
+        self.num_rounds: Optional[int] = None
+        self.adjustment_weights: Dict[str, float] = {}
+
+    def configure_fit(
+        self,
+        server_round: int,
+        parameters: Parameters,
+        client_manager: ClientManager,
+    ) -> List[Tuple[ClientProxy, FitIns]]:
+        """
+        Configure the next round of training.
+
+        Will also collect the number of rounds the training will run for in order
+        to calculate the adjustment weight step size. Fails if n_server_rounds is not set in the config
+        or if it's not an integer.
+
+        Args:
+            server_round: (int) the current server round.
+            parameters: (Parameters) the model parameters.
+            client_manager: (ClientManager) The client manager which holds all currently
+                connected clients. It must be an instance of FixedSamplingClientManager.
+
+        Returns:
+            (List[Tuple[ClientProxy, FitIns]]) the input for the clients' fit function.
+        """
+        client_fit_ins = super().configure_fit(server_round, parameters, client_manager)
+
+        assert isinstance(
+            client_manager, FixedSamplingClientManager
+        ), f"Client manager is not of type FixedSamplingClientManager: {type(client_manager)}"
+
```

Comment:
Since you have access to the `on_fit_config` here, we probably also want to confirm that `evaluate_after_fit` is true here and provide an informative error message if it is not?